java/ql/lib/semmle/code/java/dataflow/ExternalFlowConfiguration.qll

```diff
@@ -0,0 +1,95 @@
+/**
+ * This module provides extensible predicates for configuring which kinds of MaD models
+ * are applicable to a given query.
+ */
+
+private import ExternalFlowExtensions
+
+/**
+ * Holds if the specified kind of source model is supported for the current query.
+ */
+extensible private predicate supportedThreatModel(string kind);
+
+/**
+ * Holds if the specified kind of source model is containted within the specified group.
+ */
+extensible predicate threatModelGrouping(string kind, string group);
+
+/**
+ * Finds all of the threat models that are ancestors of the specified kind.
+ */
+private string parentThreatModel(string kind) {
+  exists(string parent | threatModelGrouping(kind, parent) |
+    result = parent or result = parentThreatModel(parent)
+  )
+}
+
+/**
+ * Finds all of the threat models that are descendants of the specified kind/group.
+ */
+private string childThreatModel(string group) {
+  exists(string child | threatModelGrouping(child, group) |
+    result = child or result = childThreatModel(child)
+  )
+}
+
+/**
+ * Holds if source models of the specified kind are
+ * supported for the current query.
+ */
+bindingset[kind]
+predicate supportedSourceModel(string kind) {
+  // expansive threat model includes all kinds
+  supportedThreatModel("expansive")
+  or
+  // check if this kind is supported directly
+  supportedThreatModel(kind)
+  or
+  // check if one of this kind's ancestors are supported
+  exists(string group | group = parentThreatModel(kind) | supportedThreatModel(group))
+  or
+  // if supportedThreatModel is empty, check if kind is a subtype of "standard"
+  not supportedThreatModel(_) and
+  ("standard" = parentThreatModel(kind) or "standard" = kind)
+}
+
+private string getGlobalGroups() { result = ["standard", "expansive"] }
+
+/**
+ * A class that represents a kind of any model or group.
+ */
+private class Kind extends string {
+  Kind() {
+    sourceModel(_, _, _, _, _, _, _, this, _) or
+    sinkModel(_, _, _, _, _, _, _, this, _) or
+    summaryModel(_, _, _, _, _, _, _, _, this, _) or
+    experimentalSourceModel(_, _, _, _, _, _, _, this, _, _) or
+    experimentalSinkModel(_, _, _, _, _, _, _, this, _, _) or
+    experimentalSummaryModel(_, _, _, _, _, _, _, _, this, _, _) or
+    supportedThreatModel(this) or
+    threatModelGrouping(this, _) or
+    threatModelGrouping(_, this) or
+    this = getGlobalGroups()
+  }
+}
+
+/**
+ * Gets the related source model kind(s) under the specified threat model.
+ */
+string relatedSourceModel(Kind kind) {
+  // Use the kinds provided by the query if the standard threat model is enabled or if no threat models are provided.
+  (supportedThreatModel("standard") or not supportedThreatModel(_)) and
```

Comment:
With this approach, is it possible to add new threat models to _standard_ using `threatModelGrouping`? Maybe that's not necessary since new threat models should just be explicitly used by the query, or added to the `supportedThreatModel` extensible predicate.

Reply:
No, that will be not be possible.
With this approach then basically we should consider `standard` and `expansive` reserved *group* names as they automatically mean the following
*standard*: Use exactly the source models with the source model kinds specified by the query writer.
*expansive*: Use all existing source models regardless of their kind and what the query writer specified.

It is also possible to completely override what the query writer originally intended.
There is an example of that in `test-hardcoded-sql`. Note that in the query itself we are *only* using source models *remote* kind. However, in the extensions file we specify
```yaml
extensions:

  - addsTo:
      pack: codeql/java-all
      extensible: supportedThreatModel
    data:
      - ["sql"]
```
That is, `relatedSourceModel("remote") = {"sql"}` which means that in `ExternalFlow.qll` we will only project/return the source models that has kind *sql*. (instead of the originally intended *remote*.